File: backend-spring/src/main/java/com/cronos/controller/AuthController.java
package com.cronos.controller;

import com.cronos.dto.request.LoginRequest;
import com.cronos.dto.request.RegisterRequest;
import com.cronos.dto.request.UpdateProfileRequest;
import com.cronos.dto.response.ApiResponse;
import com.cronos.security.JwtUserDetails;
import com.cronos.service.AuthService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.util.UUID;

@RestController
@RequestMapping("/api/auth")
@RequiredArgsConstructor
public class AuthController {
    private final AuthService authService;

    @PostMapping("/register")
    public ResponseEntity<?> register(@Valid @RequestBody RegisterRequest req) {
        try {
            var result = authService.register(req);
            return ResponseEntity.ok(ApiResponse.success("Registration successful", result));
        } catch (RuntimeException e) {
            return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage()));
        }
    }

    @PostMapping("/login")
    public ResponseEntity<?> login(@Valid @RequestBody LoginRequest req) {
        try {
            var result = authService.login(req);
            return ResponseEntity.ok(ApiResponse.success("Login successful", result));
        } catch (RuntimeException e) {
            return ResponseEntity.status(401).body(ApiResponse.error(e.getMessage()));
        }
    }

    @GetMapping("/profile")
    public ResponseEntity<?> getProfile(Authentication auth) {
        UUID userId = (UUID) auth.getPrincipal();
        return ResponseEntity.ok(ApiResponse.success("Profile retrieved", authService.getProfile(userId)));
    }

    @PutMapping("/profile")
    public ResponseEntity<?> updateProfile(Authentication auth, @RequestBody UpdateProfileRequest req) {
        UUID userId = (UUID) auth.getPrincipal();
        return ResponseEntity.ok(ApiResponse.success("Profile updated", authService.updateProfile(userId, req)));
    }
}

File: backend/controllers/auth_controller.go
package controllers
import (
	"net/http"
	"cronos-backend/middleware"
	"cronos-backend/models"
	"cronos-backend/services"
	"cronos-backend/utils"
	"github.com/gin-gonic/gin"
)
type AuthController struct {
	AuthService *services.AuthService
}
func NewAuthController(authService *services.AuthService) *AuthController {
	return &AuthController{AuthService: authService}
}
func (ctrl *AuthController) Register(c *gin.Context) {
	var req models.RegisterRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.BadRequest(c, "Invalid request body", err.Error())
		return
	}
	if errs := utils.ValidateStruct(req); len(errs) > 0 {
		utils.BadRequest(c, "Validation failed", errs)
		return
	}
	user, token, err := ctrl.AuthService.Register(req)
	if err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusCreated, "Registration successful", gin.H{
		"user":  user,
		"token": token,
	})
}
func (ctrl *AuthController) Login(c *gin.Context) {
	var req models.LoginRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.BadRequest(c, "Invalid request body", err.Error())
		return
	}
	if errs := utils.ValidateStruct(req); len(errs) > 0 {
		utils.BadRequest(c, "Validation failed", errs)
		return
	}
	user, token, err := ctrl.AuthService.Login(req)
	if err != nil {
		utils.Unauthorized(c, err.Error())
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Login successful", gin.H{
		"user":  user,
		"token": token,
	})
}
func (ctrl *AuthController) GetProfile(c *gin.Context) {
	userID := middleware.GetUserID(c)
	user, err := ctrl.AuthService.GetProfile(userID)
	if err != nil {
		utils.NotFound(c, err.Error())
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Profile retrieved", user)
}
func (ctrl *AuthController) UpdateProfile(c *gin.Context) {
	userID := middleware.GetUserID(c)
	var req models.UpdateProfileRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.BadRequest(c, "Invalid request body", err.Error())
		return
	}
	user, err := ctrl.AuthService.UpdateProfile(userID, req)
	if err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Profile updated", user)
}
